### PR TITLE
Start auth library

### PR DESCRIPTION
**Pull request by rogerhutchings**
Package: `lib-auth`

A library that will eventually handle auth for Panoptes projects. Currently, it's just a standalone react app that will create its own root element and mount when called - this will be the drop-in front end form side of things.

### CODE REVIEW

**Comment by rogerhutchings** [comment]
This is a working OAuth client with its own popups and session timeout monitoring. Tests and docs will be forthcoming in later PRs (I didn't want this one to get too huge)

Try it out by running `npm run dev`

File: packages/lib-auth/src/views/oauth/components/shared/Popup/Popup.spec.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    expect(wrapper.text()).to.equal(string)
+  })
+
+  describe('login button', function () {
+    it('should exist', function () {
+      const wrapper = shallow(<Popup loginFunction={spy} />)
+      expect(wrapper.find('Button').length).to.equal(1)
+    })
+
+    it('should call the login prop when clicked', function () {
```
Comment (srallen) [inline packages/lib-auth/src/views/oauth/components/shared/Popup/Popup.spec.js]: This test is failing: `TypeError: Cannot read property 'token' of undefined`
in Popup.js on line 19:3.

File: packages/lib-auth/src/views/oauth/components/TimeoutPopup/TimeoutPopup.spec.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+import React from 'react'
+
+import TimeoutPopup from './TimeoutPopup'
+
+describe('Component > OAuthApp > TimeoutPopup', function () {
+  it('should render without crashing', function () {
+    shallow(<TimeoutPopup />)
+  })
+
+  it('should derive the timeout message from the `timeLeft` prop', function () {
```
Comment (srallen) [inline packages/lib-auth/src/views/oauth/components/TimeoutPopup/TimeoutPopup.spec.js]: Test is failing. One of the expectations is coming back as false.

File: packages/lib-auth/src/views/oauth/components/TimeoutWarningPopup/TimeoutWarningPopup.spec.js [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import { shallow } from 'enzyme'
+import React from 'react'
+
+import LogoutPopup from './TimeoutWarningPopup'
```
Comment (srallen) [inline packages/lib-auth/src/views/oauth/components/TimeoutWarningPopup/TimeoutWarningPopup.spec.js]: This should be `TimeoutWarningPopup`

File: packages/lib-auth/src/views/oauth/components/shared/Popup/Popup.js [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+          align='center'
+          direction='row'
+          justify='center'
+          margin={{ top: 'medium' }}
+        >
+          <Button
+            label={counterpart('Popup.loginButton')}
+            onClick={this.login}
+            primary
+          />
```
Comment (srallen) [inline packages/lib-auth/src/views/oauth/components/shared/Popup/Popup.js]: This `Popup` component is used for all of the popup types, like the timeout warnings, so this button won't always be for login.

**Comment by srallen** [comment]
Just had a thought about the locale files, we could add a `lib-locales` library that has all of the translated strings that could be imported into various projects, like we do we the grommet theme. This would keep it all in one place for translation volunteers.

File: packages/lib-auth/README.md [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+Use this for custom projects, and anything else not running on the zooniverse.org domain.
+
+### How to Use
+
+```
+import { createOAuthClient } from '@zooniverse/auth'
+
+const auth = createOAuthClient({
+  clientId: '1234567890', // Your client ID from Doorkeeper
+  env: 'staging', // Can be `staging` or `production`. You can also specify `authorizationUri` instead if you prefer
```
[the remark this answers is not in the document]
Comment (rogerhutchings) [reply packages/lib-auth/README.md]: Ohhhh right, I see how that's confusing now. No, I meant that you can specify an `env` to determine the authorization URI (which would provide one of the Panoptes defaults), or if you want to provide a specific URI, you can use the `authorizationUri` property

**Comment by rogerhutchings** [comment]
I'll tackle more of this tomorrow, but on the topic of locale files, I want to keep them small and next to the code. Larger locale files quickly get unwieldy to edit, and hard to find errors in. Moving them into one place reduces the portability of components.

That said, we could write a couple of scripts to bundle up locale files for our translators, and split them out again afterwards. I agree that the current pattern isn't the easiest for non-coders to work with, but that's then on us to provide tools to make their lives easier, not make ours harder.

**Comment by srallen** [comment]
@rogerhutchings @eatyourgreens I'll open a proposal ADR about how to organize the translation locale files and copy in the comments from here into that.

**Comment by rogerhutchings** [comment]
Okay, I've made some changes:

> I tried the test app and the login button isn't working. There are several errors in the console...

These should all be caught now.

>  Any ideas on how this mini app will inherit the locale and color theme settings from the app it's being used with? 

For locale, I prefer having a public `setLocale` method too. For theme, I wasn't going to let devs theme it; I felt it should be a drop-in app, providing a consistent experience across all our custom front-ends. The inspiration came from libraries like [Auth0 Lock](https://auth0.com/learn/lock-demo/) and Stripe.

> My question from #104 about where to store the locale files is relevant here too. 

Nice one on setting up the ADR, we can change things round when we have a decision.

> Also not a blocker for me, but there should be a few more tests added, particularly for the functions in the controllers folder.

I'm happy to test everything, I just wanted to get a first pass for this in.

> Lastly, I find the UI a bit confusing to follow, particularly the nesting with two different HOC functions, WithPopup and WithLayer. 

I've removed them, it should be a lot clearer now; let me know what you think.

**Comment by srallen** [comment]
> For locale, I prefer having a public setLocale method too. For theme, I wasn't going to let devs theme it; I felt it should be a drop-in app, providing a consistent experience across all our custom front-ends. The inspiration came from libraries like Auth0 Lock and Stripe.

This sounds reasonable to me.

File: packages/lib-auth/src/views/oauth/components/shared/LoginButton/LoginButton.js [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+counterpart.registerTranslations('en', en)
+
+class LoginButton extends React.Component {
+  constructor () {
+    super()
+    this.login = this.login.bind(this)
+  }
+
+  login () {
+    window.open(this.props.client.token.getUri(), '_self')
```
[the remark this answers is not in the document]
Comment (rogerhutchings) [reply packages/lib-auth/src/views/oauth/components/shared/LoginButton/LoginButton.js]: via https://stackoverflow.com/questions/2587677/avoid-browser-popup-blockers: `window.open` calls are (apparently) blocked if not in response to user action, so it _should_ be okay. Regardless, I'll change this for:

```
window.location.href = this.props.client.token.getUri()
```

File: packages/lib-auth/package.json [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+    ],
+    "parser": "babel-eslint"
+  },
+  "dependencies": {
+    "client-oauth2": "^4.2.1",
+    "counterpart": "^0.18.6",
+    "lodash": "^4.17.10",
+    "luxon": "^1.3.3",
+    "query-string": "^6.1.0",
+    "react-router-dom": "^4.3.1"
```
[the remark this answers is not in the document]
Comment (rogerhutchings) [reply packages/lib-auth/package.json]: My bad - this was left in from a previous attempt to build an example app

**Comment by rogerhutchings** [comment]
Nope, that was due to me not updating the tests properly 🤦‍♂️  

- Smoke tests have been added to all the react components
- I've added the Zoo grommet theme
- I'll create issues for the remaining tasks

**Pull request merged**